handlers/soft-opt-in-consent-setter/src/main/scala/com/gu/soft_opt_in_consent_setter/SalesforceConnector.scala

```
@@ -0,0 +1,97 @@
+        decode[SFSubscription.Response](result)
+          .left
+          .map(i => SoftOptInError("SalesforceConnector", s"Could not decode SFSubscription.Response: $i. String to decode: $result"))
+      }
+  }
+
+  def getActiveSubs(identityIds: Seq[String]): Either[SoftOptInError, AssociatedSFSubscription.Response] = {
+    if (identityIds.nonEmpty)
+      doSfGetWithQuery(SfQueries.getActiveSubsQuery(identityIds))
+        .flatMap { result =>
+          decode[AssociatedSFSubscription.Response](result)
+            .left
+            .map(i => SoftOptInError("SalesforceConnector", s"Could not decode AssociatedSFSubscription.Response: $i. String to decode: $result"))
+        }
+    else
+      Right(AssociatedSFSubscription.Response(0, true, Seq[AssociatedSFSubscription.Record]()))
+  }
+
+  def doSfCompositeRequest(jsonBody: String): Either[SoftOptInError, String] = {
+    runRequest(
+      Http(s"${sfAuthDetails.instance_url}/services/data/$sfApiVersion/composite/sobjects")
+        .option(HttpOptions.readTimeout(30000))
+        .header("Authorization", s"Bearer ${sfAuthDetails.access_token}")
+        .header("Content-Type", "application/json")
+        .postData(jsonBody)
+        .method("PATCH")
+    )
+      .map(_.body)
+      .left.map(i => SoftOptInError("SalesforceConnector", s"Salesforce composite request failed: $i"))
+  }
+
+  def updateSubsInSf(updateJsonBody: String): Either[SoftOptInError, Unit] = {
+    doSfCompositeRequest(updateJsonBody)
+      .flatMap { result =>
+        decode[List[SfResponse]](result) match {
+          case Right(compositeResponse) =>
+            SfCompositeResponse(compositeResponse).errorAsString
+              .foreach(logger.warn(_))
+
+            Right(())
+          case Left(error) =>
+            Left(SoftOptInError("SalesforceConnector", s"Could not decode SfCompositeRequest.Response: $error. String to decode: $result"))
+        }
+      }
+  }
+
+}
+
+object SalesforceConnector {
```

I'm a little unsure of this pattern of an object doing auth and an instance doing the rest of the calls, why not just have one class that does both?

---

The idea was to avoid mutability in the class. To accomplish what we wanted to do (pass the config/authDetails to the class only once) we'd have to keep a variable that mutates over time. But perhaps there are better ways to do this?

---

Me and @arausuy had a chat about this and would benefit from a third opinion. @kelvin-chappell would you mind sharing your expertise on this?

---

You could move `auth` out of the companion object and into the `SalesforceConnector` class and declare it as a `lazy val` instead of a `def`.  You'd then have to have a `SalesforceConfig` as a field of the class instead of an `SfAuthDetails`.

---

It's not necessarily an improvement though as logically the auth token is a prerequisite for the connector.  So if you can't get a token you want to fail fast.